File: 7. 일 - 구현/7588/a.py
import sys

input = sys.stdin.readline

nums = []

while 1:
    n = int(input())
    if n == 0:
        break
    nums.append(n)

max_num = max(nums)
s = [0, 0] + [1] * (max_num + 1)
for i in range(2, int(max_num**0.5) + 1):
    for j in range(i, max_num // i + 1):
        s[i * j] += i
        if i != j:
            s[i * j] += j

for num in nums:
    print(f"\nAmicable numbers between 1 and {num}")
    flag = 1
    for i in range(2, num + 1):
        if s[i] != i and i < s[i] <= max_num and s[s[i]] == i:
            flag = 0
            print(f"{i} {s[i]}")
    if flag:
        print("None")

File: 7. 일 - 구현/7588/a.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
    int n,f;
    vector<int> a;
    while(1){
        cin>>n;
        if(n==0)break;
        a.push_back(n);
    }
    int m=*max_element(a.begin(),a.end());
    long long s[10000004]={0};
    for(int i=2;i<=m;i++)s[i]=1;
    for(int i=2;i<=int(sqrt(m))+1;i++)
        for(int j=i;j<=m/i+1;j++){
            s[i*j]+=i;
            if(i!=j)s[i*j]+=j;
        }
    for(int q:a){
        cout<<"\nAmicable numbers between 1 and "<<q<<"\n";
        f=1;
        for(int i=2;i<=q;i++){
            if(s[i] != i && i < s[i] && s[s[i]] == i){
                f=0;
                cout<<i<<" "<<s[i]<<"\n";
            }
        }
        if(f)cout<<"None\n";
    }
}